README.md
# Usage on BB5

## Installation

```shell
# Clone this repository
$ git clone git@bbpgitlab.epfl.ch:nse/ArchNGV.git

# Load most recent stable modules
$ module purge
$ module load unstable python/3.10.8


# Create a Python virtualenv in repository source directory
$ python -m venv /path/to/repo/.venv

# Bring the virtualenv in this shell environment
$ . /path/to/repo/.venv/bin/activate

# Install ArchNGV
$ cd /path/to/repo
$ pip install .
```

## Create circuit exemplars

```shell
# Create a directory for your circuit
$ circuit_dir=/gpfs/bbp.cscs.ch/project/projXX/$USER/ArchNGVCircuits
$ mkdir -p $circuit_dir

# Create an exemplar
$ python ./exemplar/create_exemplar.py $circuit_dir
```

## Execute cell placement

To proceed to the cell placement in one of the created exemplar:

```
# Change directory to one of the created exemplar
$ cd $circuit_dir/exemplar_ID

# Execute the "cell-placement" snakemake target
$ ./run.sh cell-placement
# -> creates file build/cell_data.h5
```

Use the `cell_data_sonata` SnakeMake task to perform output conversion to Sonata format
after the cell placement:

```shell
$ ./run.sh cell_data_sonata

# sonata file glia.h5.somata is created in the sonata.tmp directory
$ find build/sonata.tmp
build/sonata.tmp
build/sonata.tmp/nodes
build/sonata.tmp/nodes/glia.h5.somata
```

## Astrocyte Synthesis

It uses Dask for parallel calculations.

An example for a local machine usage. Pay attention that `--parallel` option is not used:
```shell
ngv synthesis \
    --config /path/to/synthesis_config.yaml' \
    --tns-distributions /path/to/tns_distributions.json \
    --tns-parameters /path/to/tns_parameters.json \
    --tns-context /path/to/tns_context.json \
    --astrocytes /path/to/glia.h5 \
    --microdomains /path/to/microdomains.h5 \
    --gliovascular-connectivity /path/to/gliovascular.h5 \
    --neuroglial-connectivity /path/to/neuroglial.h5 \
    --endfeet-areas /path/to/endfeet_areas.h5 \
    --neuronal-connectivity /path/to/edges.h5 \
    --out-morph-dir /path/to/out_morphologies
```
An example for a BB5 usage. `--parallel` option is used in conjunction with multiple exclusive
nodes, otherwise no benefits from `--parallel`:
```shell
srun -Aproj<your_project> -N2 -t=24:00:00 --exclusive \
ngv synthesis \
    --config ... \
    // all above options as before
    --parallel
```
An example of Snakemake usage. 
```shell
snakemake \
    --snakefile snakemake/Snakefile \
    --cluster-config path/to/cluster.yaml \  # see an example below
    --config bioname path/to/bioname \
    synthesis
```

Don't forget to have an entry in your `cluster.yaml` for `synthesis`
```yaml
synthesis:
  jobname: ngv_synthesis
  account: '<your project>'
  nodes: <better have at least 2, 4 is recommended>
  partition: 'prod'
  constraint: 'cpu'
  time: '04:00:00' # feel free to increase time 
  cpus-per-task: 2
  exclusive: true
  mem: 0
```

Also it is highly recommended to export in advance next Dask variables for better performance:
```shell
export DASK_DISTRIBUTED__WORKER__USE_FILE_LOCKING=False
export DASK_DISTRIBUTED__WORKER__MEMORY__TARGET=False  # don't spill to disk
export DASK_DISTRIBUTED__WORKER__MEMORY__SPILL=False  # don't spill to disk
export DASK_DISTRIBUTED__WORKER__MEMORY__PAUSE=0.80  # pause execution at 80% memory use
export DASK_DISTRIBUTED__WORKER__MEMORY__TERMINATE=0.95  # restart the worker at 95% use
# Reduce dask profile memory usage/leak (see https://github.com/dask/distributed/issues/4091)
export DASK_DISTRIBUTED__WORKER__PROFILE__INTERVAL=10000ms  # Time between statistical profiling queries
export DASK_DISTRIBUTED__WORKER__PROFILE__CYCLE=1000000ms  # Time between starting new profile
```

A final sbatch script example
```shell
#!/bin/bash
# below SBATCH options are not related to options from cluster.yaml 
#SBATCH --partition prod
#SBATCH --account proj62
#SBATCH --nodes 1  # It is not recommended to set higher because synthesis task will be launched with SBATCH options from its entry in `cluster.yaml`  
#SBATCH --time 08:00:00
#SBATCH --job-name sNGV
#SBATCH --output out-%J.log
#SBATCH --error err-%J.log
#SBATCH --mem 200000
#SBATCH --exclusive
​
export DASK_DISTRIBUTED__WORKER__USE_FILE_LOCKING=False
export DASK_DISTRIBUTED__WORKER__MEMORY__TARGET=False  # don't spill to disk
export DASK_DISTRIBUTED__WORKER__MEMORY__SPILL=False  # don't spill to disk
export DASK_DISTRIBUTED__WORKER__MEMORY__PAUSE=0.80  # pause execution at 80% memory use
export DASK_DISTRIBUTED__WORKER__MEMORY__TERMINATE=0.95  # restart the worker at 95% use
# Reduce dask profile memory usage/leak (see https://github.com/dask/distributed/issues/4091)
export DASK_DISTRIBUTED__WORKER__PROFILE__INTERVAL=10000ms  # Time between statistical profiling queries
export DASK_DISTRIBUTED__WORKER__PROFILE__CYCLE=1000000ms  # Time between starting new profile

source <venv with ArchNGV installed>/bin/activate
​
snakemake --snakefile <path to Snakefile of this project> \
          --config bioname=<path to bioname> \
          --directory <path to save results> \
          --cluster-config <path to your cluster.yaml> \
          -f synthesis
​```


CHANGELOG.rst
Changelog
=========


Version 3.1.0
-------------

Fixed
~~~~~

- The bounding box for the microdomains is now consistent with the region of interest that has been selected.
- Replaces the use of the unmaintained python package TNS by the Neurots one.



Version 3.0.3
-------------

Fixed
~~~~~

- Fixed the snakemake syntax and code style by adding snakefmt requirement and command into the lint target of tox.ini
- Move module release from 2023-03 to 2023-05 because of a libsonata issue: ImportError: cannot import name 'CircuitConfigStatus' from 'libsonata'


Version 3.0.2
-------------

Improvements
~~~~~~~~~~~~

- Limit the microdomains to the region of interest.[`BBPP152-76`_]
- thread separately the non connected component [`BBPP152-111`_]
- Switch to the new Spatial_index package version. Use multi index [`BBPP152-109`_]
- Synthesis step improvement [`BBPP152-112`_]
- Generate tetrahedral meshes [`BBPP152-113`_]
- Adapt the build microdomains step for a negative entry for the region_mask.voxel_dimension [`BBPP152-155`_]
- Add the optional use of a ROI mask [`BBPP152-176`_]


Version 2.0.2
-------------

Improvements
~~~~~~~~~~~~

- Simplified NGV exporters and removed a lot of redundant code. [`NSETM-1927`_]
- Make the NGV population types consistent with the SONATA documentation, and maintain backward
  backward compatibility with the legacy types. [`NSETM-1923`_]

Fixed
~~~~~

- Output the correct key entry 'endfeet_meshes_file' for the endfeet meshes path in the SONATA
  config. Maintain backward compatibility with the old 'endfeet_meshes' entry. [`NSETM-1904`]


Version 2.0.1
-------------

New Features
~~~~~~~~~~~~

- NGV building supports circuit-build manifests, where the ngv specific configuration has to be
  under a root entry 'ngv', [`NSETM-1872`_]

Version 2.0.0
-------------

Improvements
~~~~~~~~~~~~

- Endfeet meshes file layout has been converted to the grouped properties one. [`NSETM-1830`_]
- Merge microdomain files into a single file at the build root level. [`NSETM-1807`_]
- Use a generic file layout for grouped properties to store the microdomains that can be extended
  with and arbitrary number of additional properties. [`NSETM-1807`_]
- Multiple yaml configuration files merged into the main MANIFEST.yaml. [`NSETM-1746`_]
- Refactored the neuroglial connectivity out of ngv.py to reduce its size. [`NSETM-1768`_]
- Refactored the gliovascular connectivity out of ngv.py to reduce its size.
- The app cli has been compressed into the ngv.py module. [`NSETM-1740`_]

New Features
~~~~~~~~~~~~
- Resampling was added in synthesis to create an upper bound in morphology points [`NSETM-1778`_]. 

Removed
~~~~~~~

- SONATA config no longer stores the bioname parameters. [`NSETM-1746`_]
- The deprecated `extras` subpackage has been removed.

.. _`NSETM-1927`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1927
.. _`NSETM-1923`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1923
.. _`NSETM-1904`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1904
.. _`NSETM-1872`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1872
.. _`NSETM-1830`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1830
.. _`NSETM-1778`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1778
.. _`NSETM-1807`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1807
.. _`NSETM-1746`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1746
.. _`NSETM-1768`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1768
.. _`NSETM-1740`: https://bbpteam.epfl.ch/project/issues/browse/NSETM-1740


archngv/app/run.py
"""Run snakemake commands"""
import logging
import sys
from datetime import datetime
from pathlib import Path
import subprocess
import click
import pkg_resources
import json


L = logging.getLogger()


def _index(args, *opts):
    """Finds index position of `opts` in `args`"""
    indices = [i for i, arg in enumerate(args) if arg in opts]
    assert len(indices) < 2, f"{opts} options can't be used together, use only one of them"
    if len(indices) == 0:
        return None
    return indices[0]


def _build_args(args, bioname, modules, timestamp):
    if _index(args, '--printshellcmds', '-p') is None:
        args = ['--printshellcmds'] + args
    if _index(args, '--cores', '--jobs', '-j') is None:
        args = ['--jobs', '8'] + args
    # force the timestamp to the same value in different executions of snakemake
    args = args + ['--config', f'bioname={bioname}', f'timestamp={timestamp}']
    if modules:
        # serialize the list of strings with json to be backward compatible with Snakemake:
        # snakemake >= 5.28.0 loads config using yaml.BaseLoader,
        # snakemake < 5.28.0 loads config using eval.
        args += [f'modules={json.dumps(modules)}']
    return args


def _run_snakemake_process(cmd, errorcode=1):
    """Run the main snakemake process."""
    result = subprocess.run(cmd)
    if result.returncode != 0:
        L.error("Snakemake process failed")
        return errorcode
    return 0


def _run_summary_process(cmd, filepath: Path, errorcode=2):
    """Save the summary to file."""
    cmd = cmd + ['--detailed-summary']
    filepath.parent.mkdir(parents=True, exist_ok=True)
    with filepath.open('w') as fd:
        result = subprocess.run(cmd, stdout=fd)
    if result.returncode != 0:
        L.error("Summary process failed")
        return errorcode
    return 0


def _run_report_process(cmd, filepath: Path, errorcode=4):
    """Save the report to file."""
    cmd = cmd + ['--report', str(filepath)]
    filepath.parent.mkdir(parents=True, exist_ok=True)
    result = subprocess.run(cmd)
    if result.returncode != 0:
        L.error("Report process failed")
        return errorcode
    return 0


@cli.command(context_settings=dict(ignore_unknown_options=True, allow_extra_args=True))
@click.option(
    '-u', '--cluster-config', required=True, type=click.Path(exists=True, dir_okay=False),
    help='Path to cluster config.',
)
@click.option(
    '--bioname', required=True, type=click.Path(exists=True, file_okay=False),
    help='Path to `bioname` folder of a circuit.',
)
@click.option(
    '-m', '--module', 'modules', multiple=True, required=False,
    help='''
Modules to be overwritten. Multiple configurations are allowed, and each one
should be given in the format:\n
    module_env:module_name/module_version[,module_name/module_version...][:module_path]\n
Examples:\n
    brainbuilder:archive/2020-08,brainbuilder/0.14.0\n
    touchdetector:archive/2020-05,touchdetector/5.4.0,hpe-mpi\n
    spykfunc:archive/2020-06,spykfunc/0.15.6:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/all
    '''
)
@click.option(
    '-s', '--snakefile', required=False, type=click.Path(exists=True, dir_okay=False),
    default=pkg_resources.resource_filename(__name__, 'snakemake/Snakefile'), show_default=True,
    help='Path to workflow definition in form of a snakefile.',
)
@click.option(
    '--with-summary', is_flag=True, help='Save a summary in `logs/<timestamp>/summary.tsv`.'
)
@click.option(
    '--with-report', is_flag=True, help='Save a report in `logs/<timestamp>/report.html`.'
)
@click.pass_context
def cmd(
    ctx,
    cluster_config: str,
    bioname: str,
    modules: list,
    snakefile: str,
    with_summary: bool,
    with_report: bool,
):
    """Run a circuit-build task.

    Any additional snakemake arguments or options can be passed at the end of this command's call.
    """
    args = ctx.args
    if snakefile is None:
        snakefile = pkg_resources.resource_filename(__name__, 'snakemake/Snakefile')
    assert Path(snakefile).is_file(), f'Snakefile "{snakefile}" does not exist!'
    assert _index(args, '--config', '-C') is None, 'snakemake `--config` option is not allowed'

    timestamp = f"{datetime.now():%Y%m%dT%H%M%S}"
    args = _build_args(args, bioname, modules, timestamp)

    cmd = ['snakemake', *args, '--snakefile', snakefile, '--cluster-config', cluster_config]
    exit_code = _run_snakemake_process(cmd)

    if with_summary:

        # snakemake with the --summary/--detailed-summary option does not execute the workflow
        filepath = Path(f'logs/{timestamp}/summary.tsv')
        L.info("Creating report in %s", filepath)
        exit_code += _run_summary_process(cmd, filepath)

    if with_report:

        # snakemake with the --report option does not execute the workflow
        filepath = Path(f'logs/{timestamp}/report.html')
        L.info("Creating summary in %s", filepath)
        exit_code += _run_report_process(cmd, filepath)

    # cumulative exit code given by the union of the exit codes, only for internal use
    #   0: success
    #   1: snakemake process failed
    #   2: summary process failed
    #   4: report process failed
    sys.exit(exit_code)


@cli.command()
def snakefile_path():
    """Outputs a path to the default Snakefile."""
    click.echo(pkg_resources.resource_filename(__name__, 'snakemake/Snakefile'))

